src/stream/catch_unwind.rs

```diff
@@ -0,0 +1,47 @@
+use std::prelude::v1::*;
+use std::any::Any;
+use std::panic::{catch_unwind, UnwindSafe, AssertUnwindSafe};
+
+use super::super::{Poll, Async};
+use super::Stream;
+
+/// Stream for the `catch_unwind` combinator.
+///
+/// This is created by this `Stream::catch_unwind` method.
+#[must_use = "streams do nothing unless polled"]
+pub struct CatchUnwind<S> where S: Stream {
+    stream: S,
+}
+
+pub fn new<S>(stream: S) -> CatchUnwind<S>
+    where S: Stream + UnwindSafe,
+{
+    CatchUnwind {
+        stream: stream,
+    }
+}
+
+impl<S> Stream for CatchUnwind<S>
+    where S: Stream + UnwindSafe,
+{
+    type Item = Result<S::Item, S::Error>;
+    type Error = Box<Any + Send>;
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        let res = try!(catch_unwind(AssertUnwindSafe(|| self.stream.poll())));
```

Comment:
Ah yeah those are the semantics I'd expect at least.

On Wednesday, September 14, 2016, Stepan Koltsov <notifications@github.com>
wrote:

> In src/stream/catch_unwind.rs
> <https://github.com/alexcrichton/futures-rs/pull/151#discussion_r78697323>
> :
>
> > +pub fn new<S>(stream: S) -> CatchUnwind<S>
> > +    where S: Stream + UnwindSafe,
> > +{
> > +    CatchUnwind {
> > +        stream: stream,
> > +    }
> > +}
> > +
> > +impl<S> Stream for CatchUnwind<S>
> > +    where S: Stream + UnwindSafe,
> > +{
> > +    type Item = Result<S::Item, S::Error>;
> > +    type Error = Box<Any + Send>;
> > +
> > +    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
> > +        let res = try!(catch_unwind(AssertUnwindSafe(|| self.stream.poll())));
>
> In that case it won't be possible to recover stream after a panic, thus
> panic will be the last element of the stream (fine with me).
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/alexcrichton/futures-rs/pull/151/files/0982812b6ffd68217e0221acb1f430c6b3f61caa#r78697323>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAD95HL2KjM50ObpizL5e8DF62K1O2edks5qp6CsgaJpZM4J8SnH>
> .
>